### PR TITLE
Re-enable empty repetition near end-of-line anchor for rlike, regexp_extract and regexp_replace

### LINKED ISSUE #5659: [BUG] Minimize false positives when falling back to CPU for end of line/string anchors and newlines
**Describe the bug**
The changes in #5610 add very broad checks for patterns containing end of line anchor `$` immediately next to a newline, begin-of-line anchor `^` or repetition that could produce empty results. However, this results in false positives:
1. `$` and `^` inside character classes are being treated as anchors, eg. the pattern `0*[D$3]` has no string anchors but will still fallback to CPU.
2. We are checking the entire AST node before/after a `$`, meaning something like `\na$` is allowed, but `(\na)$` will fallback to CPU. We should instead check only the component in the group that is closest to the `$`.
3. We fallback to CPU for cases such as `^([a-z]*)([0-9]*)([a-z]*)$` which have no line terminator characters. 

**Steps/Code to reproduce bug**
Test with the above examples. 

**Expected behavior**
We should support these cases on GPU, and they should be consistent with CPU

**Environment details (please complete the following information)**
N/A

**Additional context**
None.

### PR DESCRIPTION
**Pull request by NVnavkumar**
Fixes #5659.

This enables regular expressions with empty repetitions adjacent to end-of-line anchors in every regular expression SQL function but `split()`, a use case that had been working before but had been excluded due to other edge cases that were not functioning.  This does this by modifying to checks to narrowly allow this particular case.

### CODE REVIEW

**Comment by NVnavkumar** [comment]
build

**Comment by NVnavkumar** [comment]
build

File: integration_tests/src/main/python/regexp_test.py [hunk 1]
```diff
@@ -427,21 +427,23 @@ def test_regexp_extract():
                 'regexp_extract(a, "([0-9]+)", 1)',
                 'regexp_extract(a, "([0-9])([abcd]+)", 1)',
                 'regexp_extract(a, "([0-9])([abcd]+)", 2)',
-                'regexp_extract(a, "^([a-d]*)([0-9]*)([a-d]*)\\z", 1)',
```
Comment (andygrove) [inline integration_tests/src/main/python/regexp_test.py:430]: Is there a reason we need to remove the tests with `\\z`?

Comment (NVnavkumar) [reply integration_tests/src/main/python/regexp_test.py:430]: From what I gather `\z` is disabled:

```
        case 'z' =>
          // cuDF does not support "\z" except for in split mode
          throw new RegexUnsupportedException(
            "\\z is not supported on GPU for find or replace",
            regex.position)
```

Also, this test wasn't really testing that, in PySpark, for escaping, you need to double escape. Basically it's a string inside a SQL command that needs a `\\z`. So the Python string equivalent would be `\\\\z`.

**Pull request merged**